new tabs part 1 (widgetry demo, jump-to-time, and trips table)

michaelkirk opened the pull request:
We have several "tab" like systems in the app. 

This PR works on consolidate those implementations, and incorporates some of the new designs. 

As a reminder, here's the tab spec from figma:

<img width="212" alt="Screen Shot 2021-03-18 at 10 34 39 PM" src="https://user-images.githubusercontent.com/217057/111735956-27137780-883a-11eb-9a2c-9705928393be.png">

The active tab has dark text, while the inactive tabs which you can switch to have faded text.

I chose to diverge from figma a bit, primarily because the "inactive" tabs were tough to read with the low-contrast text styling, especially on the translucent background in a way that wasn't obvious from figma.

## before

The "finished", "canceled", "unfinished" tabs:

<img width="1072" alt="Screen Shot 2021-03-18 at 10 07 54 PM" src="https://user-images.githubusercontent.com/217057/111735782-d439c000-8839-11eb-8039-9df54b5591f5.png">

The time warp:

<img width="984" alt="Screen Shot 2021-03-18 at 10 39 58 PM" src="https://user-images.githubusercontent.com/217057/111736341-e23c1080-883a-11eb-8183-119e2a07cf49.png">

## after
<img width="1198" alt="Screen Shot 2021-03-18 at 10 46 57 PM" src="https://user-images.githubusercontent.com/217057/111736878-ddc42780-883b-11eb-9564-f6b660bfb69d.png">

<img width="1198" alt="Screen Shot 2021-03-18 at 10 45 25 PM" src="https://user-images.githubusercontent.com/217057/111736731-a48bb780-883b-11eb-84a5-81d45d35db06.png">

## increasing contrast

One other thing I did, which might be controversial...

Previously the panel_bg and the panel_section were the same colors, but the panel_bg had some opacity applied. This often looked good, but if the map behind the panel has a bunch of content which is almost panel colored, then the contrast between panel_bg and panel_section is lost. 

In particular, this could make the tabs hard to differentiate. 

To address this, I doped the panel_bg to be a bit farther away from the section color. Compare the tab outline in the left image (doped) vs. right (pure).

<img width="1920" alt="Screen Shot 2021-03-18 at 10 21 25 PM" src="https://user-images.githubusercontent.com/217057/111735777-d13ecf80-8839-11eb-916c-c0ebd54a6d1f.png">

michaelkirk commented on game/src/sandbox/dashboards/misc.rs:
```diff
@@ -67,12 +67,7 @@ impl State<App> for ActiveTraffic {
-    }
-
-    fn draw(&self, g: &mut GfxCtx, app: &App) {
-        g.clear(app.cs.dialog_bg);
```
Oh, I also got rid of the green backgrounds, because it felt more cohesive. Is there a reason we should keep it?

Relatedly, I don't really understand what `draw_baselayer` is about, but once I removed the `g.clear` here, I observed some glitchy rendering until I also removed `DrawBaselayer::Custom`.

dabreegster replied:
> I also got rid of the green backgrounds, because it felt more cohesive

Agreed, still seeing the map underneath is nicer. The "fullscreen" design with the green background was just meant to draw focus away from the map and towards the data, I think, but the new design is just as easy to read, and more cohesive as you say.

> I don't really understand what draw_baselayer is about

widgetry manages a stack of states. Most of the time, the current state is totally responsible for drawing everything, including initially clearing the screen of whatever was previously drawn there. Sometimes, we want to ask the previous state to draw, then draw on top of that. Most of the time, we want to do something default (`SharedAppState::draw_default`) -- in this case, draw the map. Occasionally though, we want to avoid the (negligible) perf hit of rendering the map at all, so we use `Custom`, which means we're responsible for doing `g.clear` too.

So what you've done is go back to `DefaultDraw`, which just renders the map, then draws on top of it. Totally fine!

michaelkirk commented on game/src/sandbox/dashboards/parking_overhead.rs:
```diff
@@ -1,20 +1,104 @@
+            DashTab::ParkingOverhead.picker(ctx, app),
+            Widget::col(vec![
+                Widget::row(vec![
+                    Text::from_multiline(vec![
```
No changes to the content, other than that's it's now outlined and has "section" styling.

michaelkirk commented on game/src/sandbox/dashboards/summaries.rs:
```diff
@@ -55,14 +55,12 @@ impl TripSummaries {
                         Widget::row(vec![
-                            contingency_table(ctx, app, &filter),
-                            scatter_plot(ctx, app, &filter),
+                            contingency_table(ctx, app, &filter).bg(ctx.style().section_bg),
```
Unrelated to the tabs, but while I was here, I noticed a few things that should probably have "section" styling.

michaelkirk commented on widgetry/src/style/button_style.rs:
```diff
@@ -106,15 +106,14 @@ impl<'a, 'c> Style {
-            .image_color(solid_style.fg, ControlState::Default)
-            .image_bg_color(solid_style.bg, ControlState::Default)
-            .image_bg_color(solid_style.bg_hover, ControlState::Hovered)
+            .image_color(self.btn_solid.fg, ControlState::Default)
```
perviously we used `self.btn_tab` to style things like the icon portion of the "pop up" buttons (e.g. "switch scenario"), but the new tab styling isn't compatible with that usage, so I've introduced another button_style.

michaelkirk commented on widgetry/src/widgets/panel.rs:
```diff
@@ -437,6 +437,12 @@ impl Panel {
         }
     }
 
+    pub fn swap_contained_content(&mut self, ctx: &EventCtx, name: &str, widget: &mut Widget) {
```
Not sure about this name... maybe `swap_inner_content` would be better?

dabreegster replied:
I'm a bit confused about what this is doing. Why is the first member of the row or column special and worth swapping out?
EDIT: Ah, it's because we're assuming `name` was made from a `.container()` call and ONLY has one item. Is it worth asserting that there's only one item in there below?

API-wise: maybe lock it down to `pub(crate)` unless you have use cases outside of `TabController` in mind, and add a bit of docstring?

michaelkirk commented:
The info panel behomoth, and a few other places we do tab-ish things (challenge picker, scenario picker) are still TODO, but I wanted to have an opportunity to check in on direction before going further - though these changes _are_ intended to be mergeable as is.

dabreegster commented:
Reviewing the changes functionally (digging into the code next): this looks great! I love how the trip table now retains state as you click between different tabs. Before, if I fiddled with filters and sorting, switched tabs, then returned, everything would be reset. It's much more intuitive to remember where we left off -- nice!

I noticed a regression with buttons on the scenario picker:
![Screenshot from 2021-03-19 09-48-51](https://user-images.githubusercontent.com/1664407/111815320-c5362a80-8898-11eb-87a6-1304ceaea719.png)

> increasing contrast

This change is easier to read/distinguish, so SGTM.

dabreegster commented on game/src/sandbox/dashboards/generic_trip_table.rs:
```diff
@@ -1,114 +1,32 @@
-    }
-}
-
-impl<T: 'static, F: 'static, P: 'static + Fn(&mut EventCtx, &App, &Table<App, T, F>) -> Panel>
```
I am thrilled to see this gutted in favor of something more generic. Wooot!

dabreegster commented on game/src/sandbox/dashboards/mod.rs:
```diff
@@ -18,9 +18,7 @@ mod trip_table;
 // Oh the dashboards melted, but we still had the radio
```
You were so close to my Modest Mouse reference, but no comment about it? :P

michaelkirk replied:
All post-2000 MM references are lost on me, but I'll check it out as penance. :neckbeard:

dabreegster commented on game/src/sandbox/dashboards/parking_overhead.rs:
```diff
@@ -1,20 +1,104 @@
+
+type Transition = widgetry::Transition<App>;
 
 use crate::app::App;
```
You can pull in `crate::app::Transition` here instead of the typedef above

michaelkirk replied:
done!

dabreegster commented on game/src/sandbox/dashboards/trip_table.rs:
```diff
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
-            make_panel_finished_trips,
-        )
-    }
+pub type Transition = widgetry::Transition<App>;
```
Same -- `crate::app::Transition`

michaelkirk replied:
done!

dabreegster commented on game/src/sandbox/dashboards/trip_table.rs:
```diff
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
+            .text(format!(
+                "{} ({:.1}%) Finished Trips",
+                prettyprint_usize(finished),
+                if total > 0 {
```
Could I suggest refactoring a little helper here:

```
fn percent(x: usize, total: usize) -> f64 {
  if total > 0 {
    (x as f64) / (total as f64) * 100.0
  } else {
    0.0
  }
}
```

I halfheartedly started https://github.com/a-b-street/abstreet/blob/master/geom/src/percent.rs a while back in the hopes of cleaning up stuff like this; maybe it's worth pursuing it further later. But for this PR, at least a little helper function would be nice!

michaelkirk replied:
ha ok, but for the record this is just your code cut/pasted. 😛

dabreegster replied:
This is a classic code review technique -- make the bar unrealistically high for the next unfortunate soul who touches old code, especially when the reviewer is the original author. Sorry to subject you to this! :D

michaelkirk replied:
Done, and I also re-ordered the text. It's less "literate" but since these are effectively headers, I think it makes sense to have the static part first.

<img width="521" alt="Screen Shot 2021-03-19 at 11 25 38 AM" src="https://user-images.githubusercontent.com/217057/111826377-da658600-88a5-11eb-972d-9b9b1dbda5bf.png">

dabreegster commented on game/src/sandbox/dashboards/trip_table.rs:
```diff
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
+                } else if self.table_tabs.handle_action(ctx, &x, &mut self.panel) {
+                    // if true, tabs handled the action
+                } else {
+                    todo!("unhandled action: {}", x)
```
nit: `panic` instead of `todo`, since this should logically be unreachable in its current form, right?

michaelkirk replied:
done!

dabreegster commented on game/src/sandbox/dashboards/trip_table.rs:
```diff
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
+    fn event(&mut self, ctx: &mut EventCtx, app: &mut App) -> Transition {
+        match self.panel.event(ctx) {
+            Outcome::Clicked(x) => {
+                if self.table_tabs.active_tab_idx() == 0 && self.finished_trips_table.clicked(&x) {
```
I really like the improvements to the code here! Using `push_tab` to build them up and keeping all of the tabs "alive" at all times is quite intuitive. There's a little bit of boilerplate to associate the tab index with the content and route events, but I think it's very localized and easy to understand. The alternatives are likely much more bizarre to express in Rust, with callbacks or implementing some complicated trait. So I like this!

Looking forward to applying this to info panels in particular, there might be some more complications with conceptually keeping all tabs' content "active" at the same time. Some tabs auto-update as time passes, and this can be a bit expensive, so we wouldn't want to necessarily do it for multiple tabs. Also they fill out this weird `Details` struct to draw extra zoomed/unzoomed stuff on the map, and it wouldn't make sense to draw multiple tabs' extra content at once. Maybe we just wind up using the new tab button style there, but for the moment, leave the existing weird structures in place for updating/switching tabs.

dabreegster commented on game/src/sandbox/time_warp.rs:
```diff
@@ -225,22 +164,41 @@ impl State<App> for JumpToDelay {
+                    if self.tabs.handle_action(ctx, action, &mut self.panel) {
+                        // if true, tabs has handled the action
+                    } else {
+                        todo!("handle action: {}", action)
```
I'm guessing this was an intermediate step of developing the tab controller. Should we change this to an `assert` or something else now that this case should be unreachable?

michaelkirk replied:
done!

dabreegster commented on widgetry/src/widgets/table.rs:
```diff
@@ -136,6 +144,9 @@ impl<A, T, F> Table<A, T, F> {
             make_pagination(ctx, num_filtered, self.skip),
         ])
+        .named(&self.id)
+        // return in separate container incase caller want to apply an outter-name
```
incase -> in case
outter -> outer

michaelkirk commented:
> I noticed a regression with buttons on the scenario picker:

I agree it looks pretty bad, but to my eye, not worse than the current look:

<img width="1157" alt="Screen Shot 2021-03-19 at 10 53 41 AM" src="https://user-images.githubusercontent.com/217057/111822735-67f2a700-88a1-11eb-9b95-9ed8ca2ff29d.png">

Can you clarify what change you'd like to see here?

Is it that you want the buttons to continue to have a visible border/background rather than appearing as "plain"?

dabreegster commented:
> Is it that you want the buttons to continue to have a visible border/background rather than appearing as "plain"?

The current look definitely is jarring. But yes, just looking like plain text makes it less clear it's a button. Wouldn't a slight outline or background work?

michaelkirk commented:
> I'm a bit confused about what this is doing. Why is the first member of the row or column special and worth swapping out?

I've renamed this method and its params and added a docstring.

michaelkirk commented on game/src/info/mod.rs:
```diff
@@ -711,32 +711,16 @@ fn make_tabs(
         hyperlinks.insert(name.to_string(), link);
     }
     {
-        // stop-gap color that is semi-legible across themes until the tab redesign is completed
```
Even though I haven't put the info-tabs into the new tab system, design wise, it's already pretty close. The only visual difference would be adding the "section" background to the tab's content,  instead of a translucent background. 

That might help a little bit, but probably not much. Do you find the contrast in the "data" page to be sufficiently legible?

This is why I offset the panel_bg in https://github.com/a-b-street/abstreet/pull/571/commits/ecca6375a3b357bf1ac70dfc38138e5bd992bdc5 

I could increase that effect to increase contrast, but it's such a key part of the overall design that I'm trying to diverge only as much as necessary to get readable content.

dabreegster replied:
> Do you find the contrast in the "more data" page tabs to be sufficiently legible?

Yes, but possibly only because the tabs happen to be on top of a section with the background, so the current tab kind of merges in:
![Screenshot from 2021-03-19 11-44-04](https://user-images.githubusercontent.com/1664407/111828363-6b3d6100-88a8-11eb-8fe7-39ceff6082f0.png)

It still feels hard to distinguish for me, but I'm fine with sticking close to the design for now. Thanks!

dabreegster commented:
LGTM! Thanks for the PR. Let me know if you want to talk over how to apply this to info panels.
